avahi: set service directory and refactor module

WilliButz opened the pull request:
Re-opened PR against staging, includes changes proposed in #61945.

###### Motivation for this change
Before this PR avahi used to look for service definitions in `etc/avahi/services` in its store path.
First this leads to the situation mentioned in #60939 and it also makes adding custom services quite user-unfriendly, because they would have to be added to the build output of avahi.

###### Things done

##### Package:
For the build `AVAHI_SERVICE_DIR` is now set to `/etc/avahi/services`.

##### Module:
Types are now specified for all options.
The fixed uid and gid for the avahi user have been removed
and the avahi user is now in the avahi group.
The the generic opening of the firewall for UDP port 5353 is
now optional, but still defaults to true.

The option `extraServiceFiles` was added to specify avahi
service definitions, which are then placed in `/etc/avahi/services`.

##### Manual:
Added a section mentioning the changes for avahi.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
#### fixes #60939

/cc @Infinisil

WilliButz committed:
avahi: set AVAHI_SERVICE_DIR to well-known location

Avahi now uses `/etc/avahi/services` instead of its
store path to look for files with service definitions.

WilliButz committed:
nixos/avahi: refactor module, add option `extraServiceFiles`

Types are now specified for all options.
The fixed uid and gid for the avahi user have been removed
and the user avahi is now in the group avahi.
The the generic opening of the firewall for UDP port 5353 is
now optional, but still defaults to true.

The option `extraServiceFiles` was added to specify avahi
service definitions, which are then placed in `/etc/avahi/services`.

WilliButz committed:
nixos/ids: remove avahi uid/gid

WilliButz committed:
nixos/doc: add section about avahi changes

WilliButz commented:
@GrahamcOfBorg test avahi

peterhoeg commented:
Good initiative!

Couple of things:

a. avahi upstream provides a systemd unit file - we don't need our own service def (except for the activation)
b. Why are we removing the static uid/gid?

WilliButz commented:
@peterhoeg 
> a. avahi upstream provides a systemd unit file - we don't need our own service def (except for the activation)

Regarding `avahi-daemon.socket` the [unit file shipped with avahi](https://github.com/lathiat/avahi/blob/v0.7/avahi-daemon/avahi-daemon.socket.in) is not different from the one in nixpkgs.
Our definition for `avahi-daemon.service` has a small diff compared to [upstream](https://github.com/lathiat/avahi/blob/v0.7/avahi-daemon/avahi-daemon.service.in):
- we do not specify `ExecReload`, which I think makes sense, because we do not have our avahi config file in the build output of avahi
- we specify the extra `path` entries and `LD_LIBRARY_PATH`
- we don't speficy the unit's `Alias`

As I am not quite sure what you meant, could you maybe elaborate on why we don't need our own service definition?

> b. Why are we removing the static uid/gid?

The service [doesn't handle a lot of files](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/misc/ids.nix#L5-L7) and the ones that are handled by avahi don't rely on a static uid. Also, as the user was not even in the avahi group until now, the gid was never used - instead the group `nogroup` was.

Maybe you thought of a reason to keep the static uid?

peterhoeg commented:
> > a. avahi upstream provides a systemd unit file - we don't need our own service def (except for the activation)
> 
> As I am not quite sure what you meant, could you maybe elaborate on why we don't need our own service definition?

We can get the unit files from upstream avahi by doing this:
```
systemd.packages = with pkgs; [ avahi ];
```

And then simply override the bits that we want to change and leave the rest as-is.

> Maybe you thought of a reason to keep the static uid?

The problem is that the uid could (would) be something else on another system which isn't great. We either want to keep it the same across the board (hence the ids.nix) or ideally use ```DynamicUser = true;``` if supported to have systemd handle it for us.

fpletz commented:
> The problem is that the uid could (would) be something else on another system which isn't great.

The avahi service does not manage any persistent state so it shouldn't really matter. Why are different uids a problem?

> ideally use DynamicUser = true; if supported to have systemd handle it for us.

Avahi drops the privileges itself to a user with a username that is compiled in. It has to be run as root and the user has to exist.

Avahi does support running as non-root but the code still expects the hardcoded username to exist. That's why we can't switch to `DynamicUser`. Fixing avahi is out of scope for this PR though.

> a. avahi upstream provides a systemd unit file - we don't need our own service def (except for the activation)

The upstream systemd service unit is pretty basic and in that case I actually prefer defining the whole service in the NixOS module. When avahi will eventually support DynamicUser or other systemd shenanigans we have to touch our service definition anyway.

fpletz commented:
@worldofpeace Do you have any remarks since you requested self-review?

elseym commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,242 @@ let
-      };
+    environment.etc = (mapAttrs' (n: v: nameValuePair
+      "avahi/services/${n}.service"
+      (if builtins.substring 0 1 v == "/" then { source = v; } else { text = v; })
```
```suggestion
      { ${if types.path.check v then "source" else "text"} = v; }
```

elseym replied:
feel free to use this. it uses the existing type check from `lib.types` and only computes the attribute name instead of the entire set. :)

elseym commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,242 @@ let
+      type = types.bool;
+      default = true;
+      description = ''
+        Wether to open the firewall for UDP port 5353.
```
```suggestion
        Whether to open the firewall for UDP port 5353.
```

elseym commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,242 @@ let
+      default = null;
+      description = ''
+        List of network interfaces that should be used by the <command>avahi-daemon</command>.
+        Other interfaces will be ignored. If <literal>null</literal> all local interfaces
```
```suggestion
        Other interfaces will be ignored. If <literal>null</literal>, all local interfaces
```

elseym commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,242 @@ let
+      defaultText = literalExample "config.networking.hostName";
+      description = ''
+        Host name advertised on the LAN. If not set, avahi will use the value
+        of config.networking.hostName.
```
```suggestion
        of <option>config.networking.hostName</option>.
```

elseym reviewed (changes requested):
just a few typos :)

WilliButz commented:
I added the suggested changes and rebased on staging.

worldofpeace commented:
> @worldofpeace Do you have any remarks since you requested self-review?

Yes, it was in queue to review. I have the time now so it's incoming shortly if it's still needed.

worldofpeace reviewed (changes requested):
Just some more cleanup should be done to the options.
Feedback on the release note also.

worldofpeace commented on nixos/doc/manual/release-notes/rl-1909.xml:
```diff
@@ -162,6 +162,18 @@
+   <listitem>
+    <para>
+     The package <literal>avahi</literal> is now built using the well-known
+     location <literal>/etc/avahi/services</literal> for looking up service
```
Perhaps there should be emphasis on what actually changed, which was before it looked up service definitions from within the nix store and now looks things up properly at that location. It's now configurable to the user.

WilliButz commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,242 @@ let
+      workstation = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''Whether to register a service of type "_workstation._tcp" on the local LAN.'';
```
I personally prefer to keep the double quotes rather than escape the single quotes inside :)

elseym replied:
what do you think about keeping a consistent style by inserting line breaks right after the opening and before the closing quotes (as one would do with multiline-descriptions)?

WilliButz replied:
Sounds good :)

WilliButz commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,242 @@ let
-        description = ''
-          Extra config to append to avahi-daemon.conf.
-        '';
+      domain = mkOption {
```
Sounds like a good idea to me, though these option names come from avahi and I think it would therefore be better to keep them.
Regarding the ambiguity, I think this is not much of a problem, because the two options are on different levels.

WilliButz commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,242 @@ let
+      '';
+    };
 
+    cacheEntriesMax = mkOption {
```
Same as with the other option.

WilliButz commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,242 @@ let
+    users.users.avahi = {
+      description = "`avahi-daemon' privilege separation user";
+      home = "/var/empty";
+      group = "avahi";
```
Thank you for picking up on this :+1:

WilliButz commented:
I updated the PR to include most of the suggested changes. I'd like to wait for feedback on the renaming of the two options before adding these changes.
Also rebased on staging.

worldofpeace commented:
> I updated the PR to include most of the suggested changes. I'd like to wait for feedback on the renaming of the two options before adding these changes.
> Also rebased on staging.

The renaming of the options were essentially nitpicks and didn't really need a response.
This looks good to me :+1: thank you for being so responsive to our requests.

Lastly, it's always a good idea to actually test the service. And specifically `extraServiceFiles`.
I'll try to perform this today in a NixOS vm.

worldofpeace commented:
Actually, perhaps https://github.com/NixOS/nixpkgs/blob/bac6f67aaa8fa87425b6f5cc046534ba4b734d9c/nixos/tests/avahi.nix should be updated to test `extraServiceFiles`?

WilliButz committed:
nixos/tests/avahi: add test for extra service definitions

WilliButz commented:
@worldofpeace done. Took me a while to build staging after rebasing though :)
@GrahamcOfBorg test avahi

worldofpeace reviewed (approved):
On my side of things this looks good :cherry_blossom:

peterhoeg commented:
> The avahi service does not manage any persistent state so it shouldn't really matter. Why are different uids a problem?

It isn't a problem per se, but currently we have this nice guarantee that the ids are all the same across systems which is no longer the case for no gain.

> Avahi does support running as non-root but the code still expects the hardcoded username to exist. That's why we can't switch to `DynamicUser`. Fixing avahi is out of scope for this PR though.

Fully agree.

> > a. avahi upstream provides a systemd unit file - we don't need our own service def (except for the activation)
> 
> The upstream systemd service unit is pretty basic and in that case I actually prefer defining the whole service in the NixOS module. When avahi will eventually support DynamicUser or other systemd shenanigans we have to touch our service definition anyway.

I think as a matter of principle, we should use upstream - we really don't gain anything by maintaining our own.

flokli commented on nixos/modules/services/networking/avahi-daemon.nix:
```diff
@@ -39,215 +37,245 @@ let
-          List of non-local DNS domains to be browsed.
-        '';
-      };
+    openFirewall = mkOption {
```
We don't want to add controls to open firewall automatically:
https://github.com/NixOS/nixpkgs/pull/60124#issuecomment-485985431

elseym replied:
changing this would probably break something: https://github.com/NixOS/nixpkgs/pull/62153/files#diff-3d530481bce69e8dfb7d962446f9d8faL249

i think in this case it's a good idea to have the option to not open the port.

WilliButz replied:
Ah good, I'd prefer it to be like that as well, didn't want to change the current behaviour when refactoring though. I'll set it to false :+1:

flokli replied:
Ah, missed that part that we previously already opened the firewall unconditionally.

In that case, I guess it's fine to have that option, keep it set to true (to not break things), and allow people to set it to false and restrict the firewall.

WilliButz replied:
okay :)

worldofpeace commented:
Only comment left that may be blocking is this is a principal of whether to use `systemd.packages` or define everything manually within the module. (raised by @peterhoeg)

I generally tend to prefer if there's a service upstream that we could use, and they're practically identical...
If we were to use it we'd have to establish the alias manually though like network-manager https://github.com/NixOS/nixpkgs/blob/97aba37876c4513102b0acef05f3438c0b896840/pkgs/tools/networking/network-manager/default.nix#L102

No real loss in leaving it as is, no?

peterhoeg commented:
> I generally tend to prefer if there's a service upstream that we could use

IMHO, the best kind of code is the one you don't have to write.

globin commented:
I think this PR is definitely ready enough to be merged, any further minor issues like using the systemd unit from upstream can be addressed in a follow-up PR. This should not have to be blocked as it definitely improves the module a lot already.
